src/main/java/de/refactoringbot/services/main/FileService.java

```diff
@@ -48,63 +55,44 @@
 		}
 		return allJavaFiles;
 	}
-	
+
 	/**
-	 * This method returns all root-folders of java files (like the src folder or
-	 * the src/main/java folder from maven projects)
+	 * This method returns all root folders' absolute paths of the given java files
+	 * (like the src folder or the src/main/java folder of maven projects)
 	 * 
 	 * @param allJavaFiles
 	 * @return javaRoots
 	 * @throws FileNotFoundException
 	 */
 	public List<String> findJavaRoots(List<String> allJavaFiles) throws FileNotFoundException {
-
-		// Init roots list
-		List<String> javaRoots = new ArrayList<>();
+		Set<String> javaRoots = new HashSet<>();
 
 		for (String javaFile : allJavaFiles) {
-			// parse a file
 			FileInputStream filepath = new FileInputStream(javaFile);
-			CompilationUnit compilationUnit = LexicalPreservingPrinter.setup(JavaParser.parse(filepath));
+			CompilationUnit compilationUnit;
 
-			// Get all Classes
-			List<PackageDeclaration> packageDeclarations = compilationUnit.findAll(PackageDeclaration.class);
+			try {
+				compilationUnit = LexicalPreservingPrinter.setup(JavaParser.parse(filepath));
+			} catch (Exception e) {
+				logger.error(e.getMessage(), e);
+				continue;
+			}
 
-			// If javafile has no package
-			if (packageDeclarations.isEmpty()) {
-				// Get javafile
-				File rootlessFile = new File(javaFile);
-				// Add parent of file to root
-				if (!javaRoots.contains(rootlessFile.getParentFile().getAbsoluteFile().getAbsolutePath())) {
-					javaRoots.add(rootlessFile.getParentFile().getAbsolutePath());
-				}
-			} else {
-				// Only 1 package declaration for each file
+			File file = new File(javaFile);
+			List<PackageDeclaration> packageDeclarations = compilationUnit.findAll(PackageDeclaration.class);
+			if (!packageDeclarations.isEmpty()) {
+				// current java file should contain exactly one package declaration
 				PackageDeclaration packageDeclaration = packageDeclarations.get(0);
-				String rootPackage = null;
-
-				if (packageDeclaration.getNameAsString().split("\\.").length == 1) {
-					rootPackage = packageDeclaration.getNameAsString();
-				} else {
-					rootPackage = packageDeclaration.getNameAsString().split("\\.")[0];
-				}
-
-				// Get javafile
-				File currentFile = new File(javaFile);
+				String rootPackage = packageDeclaration.getNameAsString().split("\\.")[0];
+				String javaRoot = file.getAbsolutePath()
+						.split(Pattern.quote(File.separator) + rootPackage + Pattern.quote(File.separator))[0];
 
-				// Until finding the root package
-				while (!currentFile.isDirectory() || !currentFile.getName().equals(rootPackage)) {
-					currentFile = currentFile.getParentFile();
-				}
-
-				// Add parent of rootPackage as java root
-				if (!javaRoots.contains(currentFile.getParentFile().getAbsoluteFile().getAbsolutePath())) {
-					javaRoots.add(currentFile.getParentFile().getAbsolutePath());
-				}
+				javaRoots.add(javaRoot);
+			} else if (file.getParentFile() != null) {
```

Comment:
Could you please add a short comment here when it might be that there is no package declaration in the file and why you are accessing the parent of the file in this case?

Reply:
The case of no package declaration already gets checked earlier 

https://github.com/Refactoring-Bot/Refactoring-Bot/blob/ae7a2edccfdc23f0e328f290cc37eef586527a59/src/main/java/de/refactoringbot/services/main/FileService.java#L83

This null-check is for the case when a package is declared, but it doesn't actually exist. I ran across this case when trying to parse a sample file, which had a made-up package name. So in that case, you arrive at the root folder of the hard drive (which has no parent), meaning that the search was unsuccessful

Reply:
That line is now obsolete with your new implementation. I removed it.